Optim: APOLLO optimizer integration

zhuhanqing opened the pull request:
# What does this PR do?

This PR integrates APOLLO (Approximated Gradient Scaling for Memory-Efficient LLM Optimization) into Hugging Face's Transformers.
APOLLO is a memory-efficient optimizer designed for LLM pre-training and full-parameter fine-tuning, offering SGD-like memory cost with AdamW-level performance.

📜 Paper: https://arxiv.org/abs/2412.05270
💻 Code: https://github.com/zhuhanqing/APOLLO

## Why APOLLO?
APOLLO introduces a new level of memory efficiency for LLM optimization:
✅ Ultra-Low Memory Usage → Achieves significant savings, even beyond GaLore, approaching SGD-level efficiency.
✅ Adam(W)-Level Performance → Maintains or surpasses Adam(W) performance, validated on LLaMA models up to 7B scale.
✅ No Expensive SVD Computation → Unlike GaLore, APOLLO leverages lightweight random projection, avoiding training stalls in large-scale LLM fine-tuning.

## Third-Party Validation of APOLLO

APOLLO has been merged into [LLaMA-Factory](https://github.com/hiyouga/LLaMA-Factory/pull/6617), [FluxML](https://github.com/FluxML/Optimisers.jl/pull/196), and with a validated performance in the [post](https://bsky.app/profile/benjmurrell.bsky.social/post/3lcyfrf5b7k2u).

With these validations, merging APOLLO into Transformers would offer Hugging Face users an efficient, memory-friendly optimizer for training LLMs—reducing GPU memory requirements and making large-scale model training more accessible! 🚀

## Test of the integration

Following the approach in [PR #29588](https://github.com/huggingface/transformers/pull/29588) for GaLore, I have successfully integrated APOLLO into Hugging Face Transformers.

✅ Ensuring Correctness
To verify the integration, I have:
1️⃣ Added multiple unit tests in `tests/trainer/test_trainer.py`.
2️⃣ Manually tested the API using the following script:
```
import torch
import datasets
from transformers import TrainingArguments, AutoConfig, AutoTokenizer, AutoModelForCausalLM
import trl

train_dataset = datasets.load_dataset('imdb', split='train')

args = TrainingArguments(
    output_dir="./test-apollo",
    max_steps=100,
    per_device_train_batch_size=2,
    optim="apollo_adamw",
    optim_target_modules=["attn", "mlp"]
)

model_id = "mistralai/Mistral-7B-v0.1"

config = AutoConfig.from_pretrained(model_id)

tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_config(config).to(0)

trainer = trl.SFTTrainer(
    model=model, 
    args=args,
    train_dataset=train_dataset,
    dataset_text_field='text',
    max_seq_length=512,
)

trainer.train()
```

## Who can review?

@muellerzr and @SunMarc  – Would love your feedback on this integration! 😊
Let me know if any modifications are needed. Thanks! 🚀

zhuhanqing committed:
Added APOLLO optimizer integration

zhuhanqing commented:
Hello @muellerz, 

The code check has 3 failures as above, but it seems not from our code but from other code, 
e.g., [tests_torch](https://app.circleci.com/pipelines/github/huggingface/transformers/117737/workflows/626a7625-ef6d-4aac-8914-0b3127e9988f/jobs/1584963?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary) and [check_repository_consistency](https://app.circleci.com/jobs/github/huggingface/transformers/1584943)

SunMarc reviewed (commented):
Thanks for the PR ! Just had a look and it looks like the integration follows quite closely galore implementation. Can you simplify the code so that it is not redundant ?

zhuhanqing committed:
keep up-to-date

Merge remote-tracking branch 'upstream/main' into apollo-integration

zhuhanqing commented on docs/source/en/trainer.md:
```diff
@@ -445,6 +445,139 @@ trainer.train()
+
+model_id = "google/gemma-2b"
+
+config = AutoConfig.from_pretrained(model_id)
+
+tokenizer = AutoTokenizer.from_pretrained(model_id)
+model = AutoModelForCausalLM.from_config(config).to(0)
```
Done.

zhuhanqing commented on docs/source/en/trainer.md:
```diff
@@ -445,6 +445,139 @@ trainer.train()
+import datasets
+import trl
+
+from transformers import TrainingArguments, AutoConfig, AutoTokenizer, AutoModelForCausalLM
```
solved.

zhuhanqing committed:
fix comment

zhuhanqing commented on src/transformers/trainer.py:
```diff
@@ -1582,6 +1583,119 @@ def optimizer_hook(param):
+
+            optimizer_cls = optimizer_mapping[args.optim]
+
+            if args.optim_target_modules is None:
+                raise ValueError(
+                    "You need to define a `optim_target_modules` in order to properly use APOLLO optimizers"
+                )
```
Done!

zhuhanqing committed:
Remove redundancy: Modularize low-rank optimizer construction

zhuhanqing commented:
Hi @SunMarc and @muellerz,

I have fixed your comment and updated our commit! 
Thank you so much for your quick response and efforts in reviewing our codes.

Yep, I closely follow the GaLore implementation to add APOLLO support to ***ensure our code quality meets the standard of hugging face***
APOLLO shares a similar method to using low-rank space to reduce memory costs like GaLore, making it feasible to follow GaLore implementation. APOLLO uses the low-rank space to approximate the gradient scaling factor for full-rank raw gradient, while our previous GaLore fundamentally differs from the new APOLLO as they use a more projected gradient descent method and perform updates in low-rank space, leading to accuracy loss due to the usage of low-rank updates.

**Fix your comments**
I have fixed your comments in the new commit.

**Fix redundancy**
I believe the redundancy lies in the `src/transformers/trainer.py` since we construct the APOLLO optimizer using a similar logic as GaLore by separating modules into low-rank and non-low-rank ones and appending those params as params groups.  I modularize the construction logic such that GaLore and APOLLO only need to call a function to share the same codes.

Thank you again for your time and help!
Let me know if any modifications are needed. Thanks! 🚀

Best wishes,
Hanqing

zhuhanqing committed:
Remove redundancy: Remove useless comment

SunMarc commented on docs/source/en/trainer.md:
```diff
@@ -443,6 +443,137 @@ trainer.train()
+
+train_dataset = datasets.load_dataset('imdb', split='train')
+
+args = TrainingArguments(
+    output_dir="./test-galore",
+    max_steps=100,
+    per_device_train_batch_size=2,
+    optim="galore_adamw",
+    optim_target_modules=[r".*.attn.*", r".*.mlp.*"],
```
make sure to pass the right optim value

zhuhanqing replied:
Fixed

SunMarc commented on docs/source/en/trainer.md:
```diff
@@ -443,6 +443,137 @@ trainer.train()
+You can read more about the method in the [original repository](https://github.com/zhuhanqing/APOLLO) or the [paper](https://arxiv.org/abs/2412.05270).
+
+
+You can also perform layer-wise APOLLO by simply post-pending the optimizer name with `layerwise` like below:
```
Can you explain a bit the difference with layer-wise apollo ?

zhuhanqing replied:
Done

SunMarc commented on src/transformers/trainer.py:
```diff
@@ -1484,105 +1573,46 @@ def get_optimizer_cls_and_kwargs(
 
-                optimizer_cls = LayerWiseDummyOptimizer
-                optimizer_kwargs.update({"optimizer_dict": optimizer_dict})
+            optimizer_mapping = {
+                OptimizerNames.APOLLO_ADAMW: APOLLOAdamW,
+                OptimizerNames.APOLLO_ADAMW_LAYERWISE: APOLLOAdamW,
+            }
```
In your paper, you talk about q-apollo and apollo-mini also. Is it possible to integrate them also ?

zhuhanqing replied:
Hi @SunMarc, yeah, both APOLLO and APOLLO-Mini are supported together within the APOLLOAdamW. APOLLO-MIni needs to pass the argument to use rank-1 space and tensor-wise scaling. ***Let me add some notes in `trainer.md` to show how to use APOLLO and APOLLO-Mini.***

Q-APOLLO adopts quantization for master weights (same method as our Q-GaLore) with no modification on the optimizer part (since APOLLO is already very memory-efficient without needing extra quantization for optimizer states). The master weight quantization part cannot be easily supported in transformers now; we may consider submitting a future PR for Q-APOLLO and Q-GaLore.

SunMarc commented on src/transformers/trainer.py:
```diff
@@ -1310,6 +1311,98 @@ def get_optimizer_cls_and_kwargs(
             "eps": args.adam_epsilon,
         }
+
+        def setup_low_rank_optimizer(optimizer_name, optimizer_mapping, optim_kwargs, is_layerwise_supported=True):
+            """
+            Helper function to set up low-rank optimizers like GaLore and Apollo.
+
+            Args:
+                optimizer_name (str): Name of the optimizer.
+                optimizer_mapping (dict): Mapping of optimizer names to their classes.
+                optim_kwargs (dict): Keyword arguments for the optimizer.
+                is_layerwise_supported (bool): Whether layerwise optimization is supported.
+
+            Returns:
+                Tuple[Any, Any]: Optimizer class and updated optimizer kwargs.
+            """
```
Please add the typing for this function since you did it for the docstring.

zhuhanqing replied:
Done! Thank you for your suggestion.
```
def setup_low_rank_optimizer(
            optimizer_name: str,
            optimizer_mapping: Dict[str, Any],
            optim_kwargs: Dict[str, Any],
            is_layerwise_supported: bool = True
        ) -> Tuple[Any, Any]:
```

SunMarc commented on docs/source/en/trainer.md:
```diff
@@ -443,6 +443,137 @@ trainer.train()
+
+model_id = "google/gemma-2b"
+
+config = AutoConfig.from_pretrained(model_id)
+
+tokenizer = AutoTokenizer.from_pretrained(model_id)
+model = AutoModelForCausalLM.from_config(config).to(0)
```
still not updated, maybe you missed the commit

zhuhanqing replied:
Thank you for pointing out this! I fixed this.

SunMarc commented on docs/source/en/trainer.md:
````diff
@@ -443,6 +443,137 @@ trainer.train()
+pip install apollo-torch
+```
+
+Then simply add one of `["apollo_adamw"]` in `optim` together with `optim_target_modules`, which can be a list of strings, regex or full path corresponding to the target module names you want to adapt. Below is an end-to-end example script (make sure to `pip install trl datasets`):
+
+```python
+import torch
+import datasets
+import trl
+
+from transformers import TrainingArguments, AutoConfig, AutoTokenizer, AutoModelForCausalLM
+
+train_dataset = datasets.load_dataset('imdb', split='train')
+
+args = TrainingArguments(
+    output_dir="./test-apollo",
+    max_steps=100,
+    per_device_train_batch_size=2,
+    optim="apollo_adamw",
+    optim_target_modules=[r".*.attn.*", r".*.mlp.*"]
+)
+
+model_id = "google/gemma-2b"
+
+config = AutoConfig.from_pretrained(model_id)
````
The doc is becoming a bit bloated with three example where the only thing that changes is an argument. Maybe we can try to reduce it down to only one example and still mention the other features ?

zhuhanqing replied:
Yes, I agree that our current APOLLO description is a lazy one that simply follows the previous GaLore's. I have rewrote the description to keep it concise and clearly mention our features.

SunMarc reviewed (commented):
Thanks for iterating ! Left a couple of comments. This is much better

zhuhanqing committed:
Merge remote-tracking branch 'upstream/main' into apollo-integration

zhuhanqing committed:
Fix comment: Add typing

zhuhanqing committed:
Fix comment: Rewrite apollo desc

zhuhanqing commented:
Hi @SunMarc,

Thank so much for your feedback!

I submit two commits, with the first one adding the typing in the `setup_low_rank_optimizer` function.

For the second commit, I agree that our current description of APOLLO in `trainer.md` is lazy and simply follows the GaLore one. I rewrote the description to make sure it is clean and clear and address your comments. In particular, I show how to precisely configure APOLLO-Mini by passing the right arguments. 

Thank you again for your time and help! Looking forward to your feedback!

Best wishes,
Hanqing

SunMarc reviewed (approved):
Thanks a lot for iterating @zhuhanqing ! LGTM !

SunMarc commented:
Letting @ArthurZucker having a quick look if possible. Otherwise, I'll merge in 2/3 days !

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_36062). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

zhuhanqing commented:
Hi @SunMarc, thank you so much for your help! Btw our APOLLO has just been accepted by [MLSys 2025](https://mlsys.org/), just announced today. I am eager to see APOLLO contribute more to the open-source community, especially democratizing LLM training with the integration with HF transformer!

SunMarc commented:
Congrats for your acceptance !

ArthurZucker reviewed (commented):
Nice!!! 🚀
